### ISSUE TITLE
An error when run the solver of AMGX

### ISSUE THREAD

**Issue by nghiakvnvsd** [mfem/mfem#2027]
Hi @artv3 . I have a problem when I use AMGX for solver. My code is like:
```
...
while(...)
{
    ......
    AmgXSolver amgx;
    amgx.ReadParameters(amgx_json_file, AmgXSolver::EXTERNAL); 
    // Forms MPI teams to load balance between MPI ranks and GPUs
    amgx.InitMPITeams(MPI_COMM_WORLD, ndevices);
    amgx.SetOperator(*A.As<HypreParMatrix>());
    amgx.SetConvergenceCheck(true);
    amgx.Mult(B, X);
    amgx.Finalize();
   ....
}
...
```
It still runs normally, but at a loop it reported

```
-------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node ubuntu exited on signal 9 (Killed).
--------------------------------------------------------------------------
```
 and stop the program. I have no idea about it. May you help me? Thank you!!!

**Comment by artv3** [comment]
Hi @nghiakvnvsd how many times did it run in the loop before crashing? I'm wondering if some of the memory is not being freed correctly. Can you do a fixed number of iterations and try valgrind?

**Comment by artv3** [comment]
Also if the sparsity pattern doesn't change there is an update matrix coefficients that you can use to avoid rebuilding the amg hierarchy.

**Comment by nghiakvnvsd** [comment]
Hi @artv3 

It depends on the mesh I use. If I use the small mesh then it most likely won't happen. But when I use a large grid (486,600 elements) it happens. The number of loops that it stops is 83 but when I restart my computer, it stopped was 85. About my program I'm pretty sure I have freed correctly.

**Comment by nghiakvnvsd** [comment]
this is my code
[TO3D-modified.txt](https://github.com/mfem/mfem/files/5902743/TO3D-modified.txt)
and this is the mesh I use

https://drive.google.com/file/d/16uX5aJLYzx0aJKgzWzp_tcN53cSouqx6/view



Sorry! My code seems not good. FEM process from line 134 to 184

and I use the json file (amg_pcg.json):

```
{                                                                                                                  
    "config_version": 2,                                                                                           
    "solver": {                                                                                                    
        "preconditioner": {                                                                                        
            "print_grid_stats": 1,                                                                                 
            "print_vis_data": 0,                                                                                   
            "solver": "AMG",                                                                                       
            "smoother": {                                                                                          
                "scope": "jacobi",                                                                                 
                "solver": "JACOBI_L1",                                                                             
                "relaxation_factor": 2                                                                             
            },                                                                                                     
            "print_solve_stats": 0,                                                                                
            "presweeps": 1,                                                                                        
            "interpolator": "D2",                                                                                  
            "max_row_sum": 0.5,                                                                                    
            "strength_threshold": 0.7,                                                                             
            "coarsest_sweeps": 2,                                                                                  
            "max_iters": 1,                                                                                        
            "monitor_residual": 0,                                                                                 
            "store_res_history": 0,                                                                                
            "scope": "amg",                                                                                        
            "max_levels": 1000,                                                                                    
            "cycle": "V",                                                                                          
            "postsweeps": 1                                                                                        
        },                                                                                                         
        "solver": "PCG",                                                                                           
        "print_solve_stats": 1,                                                                                    
        "obtain_timings": 1,                                                                                       
        "max_iters": 200000,                                                                                         
        "monitor_residual": 1,                                                                                     
        "convergence": "RELATIVE_INI_CORE",                                                                        
        "scope": "main",                                                                                           
        "tolerance": 1e-7,                                                                                        
        "norm": "L2"                                                                                               
    }                                                                                                              
}     
```

I run the code by:
mpirun -np 12 TO3D-modified --amgx-file amg_pcg.json --amgx-mpi-teams -nd 3

**Comment by nghiakvnvsd** [comment]
Also, before, I have run this code using HYPRE BoomerAMG preconditioner and this is not happend like it. So I think the reason occurs when I use the AMGX for solver.

**Comment by artv3** [comment]
> Hi @artv3
> 
> It depends on the mesh I use. If I use the small mesh then it most likely won't happen. But when I use a large mesh (486,600 elements) it happens. The number of loops that it stops is 83 but when I restart my computer, it stopped was 85. About my program I'm pretty sure I have freed correctly.

To double check, with the large mesh are you are able to solve the problem at least once? If not you could be maxing out your GPU memory. Do you know how much memory the GPUs have and the storage requirements your problem needs?  Something to try is make problem a bit smaller and see if that fits on the device.

**Comment by nghiakvnvsd** [comment]
Yes, I can solve the problem at least once. This is the result when I run at once:

![image](https://user-images.githubusercontent.com/68639058/107182872-ec3f4800-6a20-11eb-9c0f-0c5a477a09c9.png)

That does seem the storage requirements my problem needs is 3.793GB?

And also. This is my GPU information:

![image](https://user-images.githubusercontent.com/68639058/107183042-48a26780-6a21-11eb-90d6-ec4317105e04.png)

I think my GPU is good for the large problem.

**Comment by artv3** [comment]
@nghiakvnvsd could you try running valgrind? 
```
 mpirun -np 12 valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose  --log-file=valgrind-out.txt TO3D-modified --amgx-file amg_pcg.json --amgx-mpi-teams -nd 3
```
and share the output?

**Comment by artv3** [comment]
@nghiakvnvsd, I think I made some progress can you give this branch a try? 
https://github.com/mfem/mfem/pull/2047  
branch name: artv3/memleak-amgX

**Comment by nghiakvnvsd** [comment]
> @nghiakvnvsd, I think I made some progress can you give this branch a try?
> #2047
> branch name: artv3/memleak-amgX

Thanks for your help. I have just tried to use this branch but the same issues occurs.

**Comment by nghiakvnvsd** [comment]
> @nghiakvnvsd, I think I made some progress can you give this branch a try?
> #2047
> branch name: artv3/memleak-amgX

In addition, not only angx but superlu has the same problem when i run for this problem. There is just boomerAMG of hypre dont have the issues.

**Comment by artv3** [comment]
> > @nghiakvnvsd, I think I made some progress can you give this branch a try?
> > #2047
> > branch name: artv3/memleak-amgX
> 
> In addition, not only angx but superlu has the same problem when i run for this problem. There is just boomerAMG of hypre dont have the issues.

I suspect it is memory leaks, if you can share the output of valgrind that would really help.

**Comment by nghiakvnvsd** [comment]
> @nghiakvnvsd could you try running valgrind on a smaller problem?
> 
> ```
>  mpirun -np 12 valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose  --log-file=valgrind-out.txt TO3D-modified --amgx-file amg_pcg.json --amgx-mpi-teams -nd 3
> ```
> 
> and share the output?

I'll try check it and send you the output. Tomorrow I will send you. Sorry because I left my lab (time at Korea)

**Comment by nghiakvnvsd** [comment]
Hi @artv3  I have just tried to run  
`mpirun -np 12 valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose  --log-file=valgrind-out.txt TO3D-modified --amgx-file amg_pcg.json --amgx-mpi-teams -nd 3`

but it returns:

```
--------------------------------------------------------------------------
mpirun was unable to find the specified executable file, and therefore
did not launch the job.  This error was first reported for process
rank 0; it may have occurred for other processes as well.

NOTE: A common cause for this error is misspelling a mpirun command
      line parameter option (remember that mpirun interprets the first
      unrecognized command line token as the executable).

Node:       ubuntu
Executable: valgrind
--------------------------------------------------------------------------
12 total processes failed to start

```

**Comment by artv3** [comment]
TO3D-modified should be the name of your executable.

**Comment by artv3** [comment]
oh wait...  hmmmm is valgrind installed?

**Comment by nghiakvnvsd** [comment]
> oh wait... hmmmm is valgrind installed?

 Sorry about that. I have just installed. It returns:
```

 valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
valgrind: TO3D-modified: command not found
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
valgrind: TO3D-modified: command not found

```

Sorry, it is first time I use c++ to run my model and this is my bad.

**Comment by artv3** [comment]
> > oh wait... hmmmm is valgrind installed?
> 
> Sorry about that. I have just installed. It returns:
> 
> ```
> 
>  valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> valgrind: TO3D-modified: command not found
> --------------------------------------------------------------------------
> Primary job  terminated normally, but 1 process returned
> a non-zero exit code. Per user-direction, the job has been aborted.
> --------------------------------------------------------------------------
> valgrind: TO3D-modified: command not found
> ```
> 
> Sorry, it is first time I use c++ to run my model and this is my bad.

try adding ``./`` before the executable i.e.  ``./TO3D-modified``

**Comment by artv3** [comment]
I also made a small change to this branch artv3/memleak-amgX, perhaps you can pull the latest changes and try it once more? Otherwise the report from valgrind is our best bet.

**Comment by nghiakvnvsd** [comment]
> I also made a small change to this branch artv3/memleak-amgX, perhaps you can pull the latest changes and try it once more? Otherwise the report from valgrind is our best bet.

it hasn't finished yet, but I can provide you with the pre-recorded results in this file:

[valgrind-out.txt](https://github.com/mfem/mfem/files/5963299/valgrind-out.txt)

**Comment by artv3** [comment]
I see some memory leaks from the MFEM/Hypre side, did the  code complete -i.e. is this the complete file? 
I suspect that these result in the code terminating early, but not a 100% sure. 

```
==7710== 16 bytes in 1 blocks are still reachable in loss record 15 of 65
==7710==    at 0x7DB2B0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7710==    by 0x1F597AF9: strdup (strdup.c:42)
==7710==    by 0x21AABA2D: opal_output_open (in /usr/local/lib/libopen-pal.so.40.20.5)
==7710==    by 0x21A8B516: opal_cuda_check_bufs (in /usr/local/lib/libopen-pal.so.40.20.5)
==7710==    by 0x2D2C1F82: ???
==7710==    by 0x1E8C8796: PMPI_Scan (in /usr/local/lib/libmpi.so.40.20.5)
==7710==    by 0x3D0FD6: mfem::ParMesh::GenerateOffsets(int, int*, mfem::Array<int>**) const (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x451F1F: mfem::ParFiniteElementSpace::GenerateGlobalOffsets() const (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x460CDC: mfem::ParFiniteElementSpace::ParInit(mfem::ParMesh*) (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x461E7B: mfem::ParFiniteElementSpace::ParFiniteElementSpace(mfem::ParMesh*, mfem::FiniteElementCollection const*, int, int) (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x28E061: main (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710== 
==7710== 16 bytes in 1 blocks are still reachable in loss record 16 of 65
==7710==    at 0x7DB4B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==7710==    by 0xA981B6: hypre_CAlloc (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0xA97DF4: hypre_handle (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0xA9009E: hypre_CSRMatrixCreate (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0xA6D747: hypre_ParCSRMatrixCreate (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x90F7AF: mfem::HypreParMatrix::HypreParMatrix(ompi_communicator_t*, int, int, int*, int*, int*, int*, int*, int*, int*, int) (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x458A09: mfem::ParFiniteElementSpace::Build_Dof_TrueDof_Matrix() const (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x459034: mfem::ParFiniteElementSpace::GetEssentialTrueDofs(mfem::Array<int> const&, mfem::Array<int>&, int) (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710==    by 0x28E11A: main (in /home/ubuntu/mfem-4.2sau/examples/amgx/TO3D-modified)
==7710== 

```

**Comment by nghiakvnvsd** [comment]
The complete file:

[valgrind-out.txt](https://github.com/mfem/mfem/files/5977154/valgrind-out.txt)

**Comment by artv3** [comment]
Update: resolved in #2047
